### PR TITLE
Makes Ratvar more powerful than the admins

### LINKED ISSUE #23528: "Dont allow singularity to be possessed" config doesnt stop admins from possessing ratvar
Probably because it's not pathed under singulo

### PR DESCRIPTION
**Pull request by silicons**
Fixes #23528

### CODE REVIEW

**Comment by KorPhaeron** [comment]
You either need to do this properly (not a chain of istypes) or we need to just give up on this config and ban admins who possess things like this if that is what the headmins want

**Comment by kevinz000** [comment]
k

File: code/modules/admin/verbs/possess.dm [hunk 1]
```diff
@@ -2,9 +2,9 @@
 	set name = "Possess Obj"
 	set category = "Object"
 
-	if(istype(O,/obj/singularity))
+	if(istype(O,/obj/singularity) || istype(O, /obj/structure/destructible/clockwork/massive/ratvar))
```
Comment (RemieRichards) [inline code/modules/admin/verbs/possess.dm]:
```
/atom/movable
    var/admin_possessable = TRUE

...

/obj/singularity
    admin_possessable = FALSE

...

/obj/structure/destructible/clockwork/massive/ratvar
    admin_possessable = FALSE

if(!admin_possessable && config.forbid_singulo_possession) //time to rename the config entry!
    ...

```

**Comment by RemieRichards** [comment]
you also need to VVlock the ```dangerous_possession``` var.

**Comment by RemieRichards** [comment]
And probably ```control_object``` too, so they can't just create their own possession via VV (This lets you skip the existing singulo possession check!!!)

**Comment by kevinz000** [comment]
Oh, wew.

**Comment by kevinz000** [comment]
I assume you meant vvlocking with vv_edit_var because vvlocked is bypassable via +DEBUG (I think?)

File: code/game/objects/objs.dm [hunk 2]
```diff
@@ -28,6 +28,10 @@
 	
 	var/dangerous_possession = FALSE	//Admin possession yes/no
 
+/obj/vv_edit_var(vname, vval)
+	if((vname == "control_object") || (vname == "dangerous_possession))
```
Comment (RemieRichards) [inline code/game/objects/objs.dm]: what is it with you and not closing strings eh?

**Comment by Jordie0608** [comment]
Alternatively a `possess_act()`; although I'm not sure it'd be better than a var since off the top of my head I can't think of any other special interactions.

Either way I think we'd probably be better off without this restriction as another "we can't prevent everything and we trust our own admins" case since there's worse ways admins can already play with black holes.

Also use a real git client you nerd.

File: code/game/objects/objs.dm (hunk 2 above)
Comment (RemieRichards) [inline code/game/objects/objs.dm]: also make it a switch.

```
switch(vname)
    if("control_object", "dangerous_possession")
        return FALSE
```

**Comment by RemieRichards** [comment]
@kevinz000 yeah.

File: code/game/objects/objs.dm (hunk 2 above)
Comment (RemieRichards) [inline code/game/objects/objs.dm]: actually, 
```
switch(vname)
    if("dangerous_possession")
        return FALSE
    if("control_object")
        if(istype(vval, /obj))
            var/obj/O = vval
            if(O.dangerous_possession) //nice try!
                return FALSE
..()
```

This lets you still VV ```control_object``` but you can't cheat possession anymore.
```



control_object

File: code/game/objects/objs.dm [hunk 3]
```diff
@@ -29,8 +29,13 @@
 	var/dangerous_possession = FALSE	//Admin possession yes/no
 
 /obj/vv_edit_var(vname, vval)
-	if((vname == "control_object") || (vname == "dangerous_possession))
-		return FALSE
+	switch(vname)
+		if("dangerous_possession")
+			return FALSE
+		if("control_object")
+			var/obj/O = vval)
```
Comment (RemieRichards) [inline code/game/objects/objs.dm]: ```)```

**Comment by RemieRichards** [comment]
This PR is good now.

**Comment by kevinz000** [comment]
:3c

**Comment by coiax** [comment]
Why do we need this

**Comment by kevinz000** [comment]
@coiax it's a bug on the bug tracker and i want to go back to ruining things with ruined features so i'm going to fix it?

**Comment by Thunder12345** [comment]
Yeah, probably better to just do away with blocking singulo possession entirely, and deal with people that abuse it. It doesn't actually stop anything because you can just buildmode throw it anyway.

**Comment by RemieRichards** [comment]
Thank you for reporting that bug Thunder.

**Comment by Thunder12345** [comment]
But then how will I launch a supermatter singulo around CentComm at round end

**Comment by RemieRichards** [comment]
you wont.

>Thank you for reporting that bug Thunder.

**Comment by Thunder12345** [comment]
Do we really need to specifically block admins from doing something everyone knows is for round end shittery, at the very most? We just say "no fucking with megafauna", and I don't remember anyone abusing megafauna. Not every issue needs a code solution.

**Comment by Jordie0608** [comment]
It's not feasible to prevent all the ways in which things can be abused without using spaghetti code. For a particularly vermicelli example I tried just now:
- Spawn singulo beacon
- Mark a live powernet and add to beacon
- Mark singulo beacon and add to victim's contents

**Comment by RemieRichards** [comment]
simple, now you can't VV singulo beacons.

**Comment by coiax** [comment]
I think after a while, you'll have to accept that a malicious admin can do bad things. We should just log the shit out of it, and ban them afterwards.

**Comment by RemieRichards** [comment]
Yes, ban them... with code!

**Comment by kevinz000** [comment]
leave me alone im just farming fixes :3c

**Comment by ExcessiveUseOfCobblestone** [comment]
If an admin wants to grief there's a lot more ways they can do it better than spawning the justicar or moving a singularity...

I never understood the megafauna meme when we specifically put in code that allowed admins to spawn "gimped" [medal-less] versions of them IIRC that supposedly couldn't be changed, but oh well.

**Comment by Anonmare** [comment]
It's not a bug, it's a feature

**Comment by Cheridan** [comment]
You can just var-edit the singularity's target to your ghost, and then hover over people who you want to get eaten. There's so many ways to bypass these restrictions that they're pointless. This is an issue for headmins and other server administrators, not for code.

**Comment by Cyberboss** [comment]
None of you have even mentioned the glaring hole which gets around all possession bans. Also it's glaringly obvious when a sing is being possessed and I'd expect players to report it if they got killed in round by one. Honestly, the whole lockdown doesn't really make sense when almost any admin can do so much worse.

**Comment by Iamgoofball** [comment]
@RemieRichards 
Thunder has previously with-held information used to bypass shit like this, and abuses it himself. He also has refused to disclose methods to bypass admin blocks in the past.

**Comment by Thunder12345** [comment]
@Iamgoofball 
I'm sorry what, point to when this happened

**Comment by Dorsisdwarf** [comment]
At some point you have to realise that if an admin wants to grief they can just mass-qdel all humans. Direct possession blocks are a good idea to make it obvious that @this is shit you aren't allowed to do@ to particularly dim admins.

**Comment by Cyberboss** [comment]
OK to reiterate, I don't mind the possession blocks. But trying to block every possible method is an exercise in futility

**Comment by PKPenguin321** [comment]
I've been reading the points in this thread and I think I actually agree. No point in restricting this in the code, maybe just nix this config option.

**Comment by Iamgoofball** [comment]
I'll need to check back but I tried to add a feature that had an admin lock and you threatened to bypass it entirely

**Comment by Thunder12345** [comment]
Was this the time you threatened to make cat people donator only, and I predicted people would just get purrbation given out, same as when it was disabled for a while?

**Comment by Iamgoofball** [comment]
Yup, that was it. You refused to comply with locks to prevent admins from handing out donator benefits to non-paying players.

**Comment by Cyberboss** [comment]
>Purrbation

goof pls

**Comment by Iamgoofball** [comment]
i didnt name it that retarded shit

**Comment by jughu** [comment]
>nerfing admins
🤔

**Comment by kevinz000** [comment]
If you guys want to axe the possession business just close the PR and open one to remove the checks and option, just saying, only wanted to get one issue out of the way :P @PKPenguin321 @ChangelingRain @whatever ausops is here

**Comment by optimumtact** [comment]
either we take away literally every single advanced admin function like viewvar or sdql or calling procs or even buildmode and or spawning mobs/people/items or weapons or we accept that the best that can be done to prevent admins doing things like this is enforcement from other admins.

**Comment by optimumtact** [comment]
also it's @ausops

**Comment by PKPenguin321** [comment]
I would Kevin but I don't have maintainer here

File: code/modules/admin/verbs/possess.dm [hunk 4]
```diff
@@ -2,10 +2,9 @@
 	set name = "Possess Obj"
 	set category = "Object"
 
-	if(istype(O,/obj/singularity))
-		if(config.forbid_singulo_possession)
-			to_chat(usr, "It is forbidden to possess singularities.")
-			return
+	if(!O.dangerous_possession && config.forbid_singulo_possession)
```
Comment (lzimann) [inline code/modules/admin/verbs/possess.dm]: this will make you fail to possess every single item, except the ones you are trying to prevent

Comment (coiax) [inline code/modules/admin/verbs/possess.dm]: ADMINS MUST ONLY POSSESS RATVAR AND THE SINGULO

**Comment by KorPhaeron** [comment]
The singulo possession thing is nice not because its useful code but because it sends a pretty clear message that admins arent supposed to do it and you can punish people who willfully bypass it.

Otherwise it would fall into the same subjective grey area of "events" that nobody wants to touch

**Comment by coiax** [comment]
It sends a pretty clear message by being an unused option in the config file?

**Comment by KorPhaeron** [comment]
Its not unused, its enabled on our server.

Or it was for most of the last 7 years anyway

**Comment by coiax** [comment]
And yet admins can still spawn 10 tesla balls. Or they just use the throwing skill they have?

Admins have unlimited cosmic power, and we're basically marking off a tiny corner of a sandpit and preventing them from entering it. But they can still just pick up the ENTIRE SANDPIT and beat people to death with it.

File: code/modules/admin/verbs/possess.dm (hunk 4 above)
Comment (kevinz000) [inline code/modules/admin/verbs/possess.dm]: FUCK

**Comment by kevinz000** [comment]
if anything this sends a message about not being a shitter and using your powers to grief for fun.

**Comment by lzimann** [comment]
There are many way to use powers to grief, and many are way more powerful than possessing a singulo or ravtar. There is no way to prevent those things from happening without completely limiting what an admin can actually do. This whole thing is more of a server-side problem in how they want to handle the admins abusing powers.

**Comment by AnturK** [comment]
I don't believe we need admin restrictions other than preventing easy ways of crashing the whole server. Admin oversight is not something that should be done by code besides logging. That said this is just fixing oversight in current system jeez.

**Pull request merged**